### PR TITLE
firefox: enable darwin builds; also some of its derivatives (thunderbird, librewolf, floorp)

### LINKED ISSUE #353058: Package request: firefox-devedition
**Project description**

The package already exists (https://search.nixos.org/packages?channel=unstable&show=firefox-devedition) but Darwin is missing

**Metadata**

* homepage URL: https://www.mozilla.org/en-US/firefox/developer/
* source URL:
* license: 
* platforms: darwin

https://download.mozilla.org/?product=firefox-devedition-latest-ssl&os=osx&lang=en-US

### PR DESCRIPTION
**Pull request by booxter**
Enable Darwin builds for firefox (unwrapped).

The builds use vendored libraries. Using nixpkgs libraries is left for a later exercise (it may require build system patches due to insistence of the upstream on bundling for MacOS).

I'm adding myself as a maintainer for the package to reflect that I'm willing to help maintaining the Darwin integration of the package going forward.

The PR is currently based on staging-next that contains the new Apple SDK this change assumes.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [X] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by emilazy — COMMENTED** [review]
I skimmed the code while this was being worked on and it seems good. After discussion with @mweinelt we concluded that this should wait until after the 24.11 branch‐off as a “trial run”, as we want to make sure Darwin support for this package will be sustainable and that it won’t get in the way of shipping security updates to Linux users.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 1]
```diff
@@ -258,11 +265,16 @@ buildStdenv.mkDerivation {
   postPatch = ''
     rm -rf obj-x86_64-pc-linux-gnu
     patchShebangs mach build
+  '' + lib.optionalString stdenv.hostPlatform.isDarwin ''
+    # build system doesn't honor BINDGEN_EXTRA_CLANG_ARGS at all; enforce access to SDK
+    substituteInPlace config/makefiles/rust.mk --replace-fail \
+      'export BINDGEN_EXTRA_CLANG_ARGS:=$(filter --target=%,$(BINDGEN_SYSTEM_FLAGS))' \
+      'export BINDGEN_EXTRA_CLANG_ARGS:=$(filter --target=%,$(BINDGEN_SYSTEM_FLAGS)) -F${apple-sdk_14.sdkroot}/System/Library/Frameworks'
```
[the remark this answers is not in the document]
Comment (booxter) [reply pkgs/applications/networking/browsers/firefox/common.nix]: I'll leave a note to clean it up whenever it's in master.

**Comment by booxter** [comment]
This should be ready for merge. I am using this package for some time with no obvious issues on Darwin. (Also the Thunderbird package that is a derivative.) We are early in release cycle and I think it's a good time to land it.

**Comment by emilazy** [comment]
24.11 doesn’t even branch off for two more weeks :) Once it does, my latest `staging` changes will be in, so it’ll be possible to merge this into `master` without the `-F` hack.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 2]
```diff
@@ -259,11 +266,17 @@ buildStdenv.mkDerivation {
   postPatch = ''
     rm -rf obj-x86_64-pc-linux-gnu
     patchShebangs mach build
+  '' + lib.optionalString stdenv.hostPlatform.isDarwin ''
+    # build system doesn't honor BINDGEN_EXTRA_CLANG_ARGS at all; enforce access to SDK
+    # Consider removing after https://github.com/NixOS/nixpkgs/pull/349555 lands in master
```
Comment (hraban) [inline pkgs/applications/networking/browsers/firefox/common.nix:271]: 349555 has landed in master https://nixpk.gs/pr-tracker.html?pr=349555

**Review by hraban — APPROVED** [review]
Lgtm. de7585dec87f2446b02642df347cd3c5c6208eeb builds and works here (aarch64-darwin). Thanks 👍

**Review by emilazy — APPROVED** [review]
Looks good to me. Will leave it to @mweinelt for final sign‐off.

**Comment by Eveeifyeve** [comment]
I mentioned this in zen broweser as it will have to add support for it.

**Review by mweinelt — APPROVED** [review]

**Pull request merged**